### ISSUE TITLE
Typing of decorators that take an `Enum`

### ISSUE THREAD

**Issue by magnunm** [microsoft/pyright#2889]
**Is your feature request related to a problem? Please describe.**
I am having a problem typing decorators that take an `Enum` class. Here is a minimal example:
```python
from enum import Enum
from typing import Type


def enum_decorator(enumeration: Type[Enum]):
    return enumeration


@enum_decorator
class Foo(Enum):
    A = "a"


print(Foo.A)  # Pyright: Cannot access member "A" for type "Type[Enum]"
```
If I remove the type annotation `Type[Enum]` in the decorator, there is no error. And even go-to-definition from `Foo.A` to `A` in `Foo` works without `Type[Enum]`.

**Describe the solution you'd like**
I would like to be able to type the enumeration decorator. `mypy` for example does not complain about the above snippet. If there is some other way to type it than `Type[Enum]` that works with Pyright I am happy to use that.

**Comment by JelleZijlstra** [comment]
Does it work if you use a TypeVar bound to Type[Enum]?

**Comment by erictraut** [comment]
You haven't provided a return type annotation for your `enum_decorator`, so pyright is inferring its type based on the return expressions(s). In this case, the inferred return type is `Type[Enum]`. So the result of your decorator is that the exact enum class is lost and is replaced with its base class `Type[Enum]`. If you want to retain the original class, you need to use a `TypeVar` like Jelle suggested.

```python
_T = TypeVar("_T", bound=Enum)

def enum_decorator(enumeration: Type[_T]) -> Type[_T]:
    return enumeration
```
